src/auth/index.js

```diff
@@ -178,103 +213,121 @@ export default class Auth {
+    return this.storage.get(authCommon.IMPERSONATION_USER_KEY).then((userId) =>
+      client._do(`/admin/users/${userId}/impersonate`, 'POST', { refreshOnFailure: false, useRefreshToken: true })
+    )
       .then(response => response.json())
       .then(json => this.set(json))
       .catch(e => {
-        this.stopImpersonation();
-        throw e;  // rethrow
+        return this.stopImpersonation().then(() => {
+          throw e; // rethrow
+        });
       });
   }
 
   startImpersonation(client, userId) {
-    if (!this.authedId()) {
-      return Promise.reject(new StitchError('Must auth first'));
-    }
-
-    if (this.isImpersonatingUser()) {
-      return Promise.reject(new StitchError('Already impersonating a user'));
-    }
-
-    this.storage.set(authCommon.IMPERSONATION_ACTIVE_KEY, 'true');
-    this.storage.set(authCommon.IMPERSONATION_USER_KEY, userId);
+    return this.authedId().then((authedId) => {
+      if (!authedId) {
+        Promise.reject(new StitchError('Must auth first'));
+      }
+      return this.isImpersonatingUser();
+    }).then((isImpersonatingUser) => {
+      if (isImpersonatingUser) {
+        return Promise.reject(new StitchError('Already impersonating a user'));
+      }
 
-    let realUserAuth = JSON.parse(this.storage.get(authCommon.USER_AUTH_KEY));
-    this.storage.set(authCommon.IMPERSONATION_REAL_USER_AUTH_KEY, JSON.stringify(realUserAuth));
-    return this.refreshImpersonation(client);
+      return this.storage.set(authCommon.IMPERSONATION_ACTIVE_KEY, 'true');
+    }).then(() =>
+      this.storage.set(authCommon.IMPERSONATION_USER_KEY, userId)
+    ).then(() =>
+      this.storage.get(authCommon.USER_AUTH_KEY)
+    ).then((userAuth) => {
+      let realUserAuth = JSON.parse(userAuth);
+      return this.storage.set(authCommon.IMPERSONATION_REAL_USER_AUTH_KEY, JSON.stringify(realUserAuth));
+    }).then(() => this.refreshImpersonation(client));
   }
 
   stopImpersonation() {
-    if (!this.isImpersonatingUser()) {
-      throw new StitchError('Not impersonating a user');
-    }
+    return this.isImpersonatingUser().then((isImpersonatingUser) => {
+      if (!isImpersonatingUser) {
+        throw new StitchError('Not impersonating a user');
+      }
 
-    return new Promise((resolve, reject) => {
-      let realUserAuth = JSON.parse(this.storage.get(authCommon.IMPERSONATION_REAL_USER_AUTH_KEY));
-      this.set(realUserAuth);
-      this.clearImpersonation();
-      resolve();
-    });
+      return this.storage.get(authCommon.IMPERSONATION_REAL_USER_AUTH_KEY);
+    }).then((userAuth) => {
+      let realUserAuth = JSON.parse(userAuth);
+      return this.set(realUserAuth);
+    }).then(() => this.clearImpersonation());
   }
 
   clearImpersonation() {
-    this.storage.remove(authCommon.IMPERSONATION_ACTIVE_KEY);
-    this.storage.remove(authCommon.IMPERSONATION_USER_KEY);
-    this.storage.remove(authCommon.IMPERSONATION_REAL_USER_AUTH_KEY);
+    Promise.all([
```

Comment:
nit: looks like this was overlooked in the original code but I think we should just `return` this promise to the caller so that it functions properly in a promise chain